File: fourinaline/fourinaline.lua
function fsize (file)
  local current = file:seek()      -- get current position
  local size = file:seek("end")    -- get file size
  file:seek("set", current)        -- restore position
  return size
end
function min (a, b)
  if a < b then return a else return b end
end

local resultString = {"1","2","draw","play"}
local returnCode = 3
local f = io.input(arg[1])
local readOffset = fsize(f)
local BUFFER_SZ = 4096
local numReadCycles = math.floor(readOffset / BUFFER_SZ)
f:seek("set", readOffset - min(BUFFER_SZ, readOffset))
local currentRowIdx = 2
local rdBuf = f:read(min(BUFFER_SZ, readOffset))
local rdPtr = string.len(rdBuf) - 1
local startPtr = string.find(rdBuf, "\n")
local cols = string.find(string.sub(rdBuf, startPtr + 1), "\n") / 2
readOffset = readOffset - min(BUFFER_SZ, readOffset)
local ew, sn, swne, senw = {}, {}, {}, {}
senw[1] = {}
senw[2] = {}
for idx=1,cols+1 do
    ew[idx] = 0
    sn[idx] = 0
    swne[idx] = 0
    senw[1][idx] = 0
    senw[2][idx] = 0
end

repeat
	while rdPtr > startPtr do
		currentRowIdx = 3 - currentRowIdx
		local i = cols + 1
		local value = string.byte(rdBuf, rdPtr) - 48
		ew[i] = value
		sn[i] = value
		swne[i] = value
		senw[currentRowIdx][i] = value

		while i ~= 2 do
			i = i - 1
			rdPtr = rdPtr - 2
			value = string.byte(rdBuf, rdPtr) - 48
			if value > 0 then
				ew[i] = (ew[i + 1] * 4) + value
				sn[i] = (sn[i] * 4) + value
				swne[i] = (swne[i - 1] * 4) + value
				senw[currentRowIdx][i] = (senw[3-currentRowIdx][i + 1] * 4) + value
				if ew[i] == 0x55 or ew[i] == 0xaa or sn[i] == 0x55 or sn[i] == 0xaa or swne[i] == 0x55 or swne[i] == 0xaa or senw[currentRowIdx][i] == 0x55 or senw[currentRowIdx][i] == 0xaa then 
					print(resultString[value])
					os.exit(0)
				end
			else
				ew[i] = 0
				sn[i] = 0
				swne[i] = 0
				senw[currentRowIdx][i] = 0;
				returnCode = 4
			end
		end
		rdPtr = rdPtr - 2
	end
	readOffset = readOffset - min(BUFFER_SZ - startPtr, readOffset)
	f:seek("set", readOffset)
	local tmpStr = string.sub(rdBuf, 1, startPtr)
	rdBuf = f:read(min(BUFFER_SZ - startPtr, readOffset)) .. tmpStr
	rdPtr = string.len(rdBuf) - 1
	startPtr = string.find(rdBuf, "\n")
	numReadCycles = numReadCycles - 1
until numReadCycles ~= 0
print(resultString[returnCode])

File: fourinaline/fourinaline.cpp
#include <stdio.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>

#define likely(x) __builtin_expect((x), 1)
#define unlikely(x) __builtin_expect((x), 0)
#define min(a, b) (((a) < (b)) ? (a) : (b))
#define BUFFER_SZ (4096 * 1)
 
static const char * const resultString[] = { 0, "1\n", "2\n", "draw\n", "play\n" };
int main(int z, char **v)
{
	char rdBuf[BUFFER_SZ + 1], returnCode = 3;
	int fd = open(v[1], O_RDONLY);
	posix_fadvise(fd, 0, 0, POSIX_FADV_NOREUSE);
	struct stat fstatData; fstat(fd, &fstatData);
	unsigned int numReadCycles = fstatData.st_size / BUFFER_SZ;
	register off_t readOffset = fstatData.st_size;
	unsigned int rdBytes = pread(fd, rdBuf, min(BUFFER_SZ, readOffset), readOffset - min(BUFFER_SZ, readOffset)); readOffset -= min(BUFFER_SZ, readOffset);
	register char *rdPtr = rdBuf, *startPtr = strchr(rdBuf, '\n');
	register unsigned int cols = (strchr(startPtr + 1, '\n') - startPtr) / 2;
	register unsigned int currentRowIdx = 1;
	unsigned char ew[cols + 1], sn[cols + 1], swne[cols + 1], senw[2][cols + 1];
	register unsigned char *_ew, *_sn, *_senw, *_swne; swne[0]= 0;
	
	// Read upto BUFFER_SZ bytes in cycles
	do {
		rdPtr += rdBytes;
		while (rdPtr > startPtr) {
			currentRowIdx = !currentRowIdx;
			_ew = &ew[cols]; _sn = &sn[cols]; _swne = &swne[cols]; _senw = &senw[currentRowIdx][cols];
			*_ew = *_sn = *_swne = *_senw = (*rdPtr - '0');
			for (register int i = cols - 1; i; --i) {
				--_ew; --_sn; --_swne; --_senw; rdPtr -= 2;
				const register unsigned char value = *rdPtr - '0';
				if (value) {
					*_ew = ((*(_ew + 1)) << 2) | value;
					*_sn = (*_sn << 2) | value;
					*_swne = (*(_swne - 1) << 2) | value;
					*_senw = (senw[!currentRowIdx][i + 1] << 2) | value;
					if ((*_ew == 0x55) | (*_ew == 0xaa) | (*_sn == 0x55) | (*_sn == 0xaa) | (*_swne == 0x55) | (*_swne == 0xaa) | (*_senw == 0x55) | (*_senw == 0xaa)) { write(1, resultString[value], 2); _exit(0); }
				} else { *_ew = *_sn = *_swne = *_senw = 0; returnCode =  4; }
			}
			rdPtr -= 2;
		}
		rdBytes = startPtr - rdBuf;
		if (readOffset - min(BUFFER_SZ - rdBytes, readOffset))
			memmove(rdBuf + BUFFER_SZ - rdBytes, rdBuf, rdBytes);
		else
			memmove(rdBuf + readOffset, rdBuf, rdBytes);
		rdPtr = rdBuf + rdBytes;
		readOffset -= min(BUFFER_SZ - rdBytes, readOffset);
		rdBytes = pread(fd, rdBuf, min(BUFFER_SZ - rdBytes, readOffset), readOffset);
		startPtr = strchr(rdBuf, '\n');
	} while (numReadCycles--);
	write(1, resultString[returnCode], 5); _exit(0);
}
